### PR TITLE
Add edge to Entity when creating Workflows.

### PR DESCRIPTION
**Pull request by tedsharpe**

### CODE REVIEW

**Comment by dvoet** [comment]
looks good to me, @abaumann this modifies tests you wanna wait for your PR?

**Comment by abaumann** [comment]
sorry i missed this before, i have my PR up also, so either way, im happy to fix it on my end rather than the other way around if you want to merge this and ill rebase (again)

**Comment by dvoet** [comment]
this needs a rebase

**Comment by abaumann** [comment]
godspeed

**Comment by tedsharpe** [comment]
This has been rebased and is ready to review and merge.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala [hunk 1]
```diff
@@ -61,12 +61,15 @@ class GraphSubmissionDAO extends SubmissionDAO with GraphDAO {
...
     txn withGraph { db =>
-      val workspace = getWorkspaceVertex(db, workspaceNamespace, workspaceName).getOrElse(throw new IllegalArgumentException(s"workspace ${workspaceNamespace}/${workspaceName} does not exist"))
+      val workspaceVertex = getWorkspaceVertex(db, workspaceNamespace, workspaceName).getOrElse(throw new IllegalArgumentException(s"workspace ${workspaceNamespace}/${workspaceName} does not exist"))
       val submissionVertex = setVertexProperties(submission, addVertex(db, null))
-      workspace.addEdge(ExecutionEdgeTypes.submissionEdgeType, submissionVertex)
+      workspaceVertex.addEdge(ExecutionEdgeTypes.submissionEdgeType, submissionVertex)
       submission.workflow.foreach { workflow =>
         val workflowVertex = setVertexProperties(workflow, addVertex(db, null))
-        workspace.addEdge(ExecutionEdgeTypes.workflowEdgeType, workflowVertex)
+        val entityVertex =  getSinglePipelineResult(new GremlinPipeline(workspaceVertex).out(submission.entityType).filter(hasProperty("_name", workflow.entityName))).
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala]: sorry that I missed this before: the submission entity type is not necessarily the same as the workflow entity type. A submission's entity type will usually be a set and the workflow's entity type will be a member of that set. The workflow case class needs to include an entity type.

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala]: we should also have an edge from the submission to the entity the submission was started on

**Comment by dvoet** [comment]
@abaumann this is ready

File: src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 2]
```diff
@@ -61,7 +61,7 @@ object Boot extends App {
                     "rawls-service")
 
     system.actorOf(SubmissionSupervisor.props(
-      new GraphSubmissionDAO(),
+      new GraphSubmissionDAO(new GraphWorkflowDAO()),
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: no change needed, but is there a reason we would need to pass a DAO to another DAO in a constructor - can't a DAO just instantiate one as needed - they aren't tied to anything, the functions themselves use transactions

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala [hunk 3]
```diff
@@ -66,12 +79,19 @@ class GraphSubmissionDAO extends SubmissionDAO with GraphDAO {
...
+        getOrElse(throw new IllegalArgumentException(s"entity of type ${submission.entityType} named ${submission.entityName} does not exist in workspace ${workspaceNamespace}/${workspaceName}"))
+      submissionVertex.addEdge(ExecutionEdgeTypes.runOnEdgeType, submissionEntityVertex)
+      workspaceVertex.addEdge(ExecutionEdgeTypes.submissionEdgeType, submissionVertex)
+
       submission.workflow.foreach { workflow =>
         val workflowVertex = setVertexProperties(workflow, addVertex(db, null))
-        workspace.addEdge(ExecutionEdgeTypes.workflowEdgeType, workflowVertex)
+        val entityVertex =  getSinglePipelineResult(new GremlinPipeline(workspaceVertex).out(workflow.entityType).filter(hasProperty("_name", workflow.entityName))).
+                               getOrElse(throw new IllegalArgumentException(s"entity of type ${workflow.entityType} named ${workflow.entityName} does not exist in workspace ${workspaceNamespace}/${workspaceName}"))
+        workflowVertex.addEdge(ExecutionEdgeTypes.runOnEdgeType, entityVertex)
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala]: So the edge goes from workflow to entity, not entity to workflow, we would just need to follow in going edges from entities to find associated workflows I guess

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala [hunk 4]
```diff
@@ -101,10 +121,20 @@ class GraphSubmissionDAO extends SubmissionDAO with GraphDAO {
...
-    val fullMap = propertiesMap ++ Map("workspaceNamespace"->workspaceNamespace,"workspaceName"->workspaceName,"workflow"->workflows.toSeq)
-    fromPropertyMap[Submission](fullMap)
+    val entityVertex = {
+      val vertexes = vertex.getVertices(Direction.OUT, ExecutionEdgeTypes.runOnEdgeType).toList
+      if (vertexes.size != 1) {
+        throw new RawlsException(s"submission with vertex ${vertex} does not have 1 and only 1 entity")
+      }
+      vertexes.head
+    }
+    val workflows = new GremlinPipeline(vertex).out(ExecutionEdgeTypes.workflowEdgeType).toList map { graphWorkflowDAO.fromVertex(_, workspaceNamespace, workspaceName) }
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala]: isn't this every workflow, which could come from more than one submission?

File: src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala (hunk 2 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: meh... just in the spirit of declaring dependencies but not a big deal

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala (hunk 3 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphSubmissionDAO.scala]: correct, right now the stories read more like show what entity a workflow is associated to and not vice versa but does directionality actually make a difference?

**Pull request merged**